cpp/include/raft/util/bitset.cuh

```diff
@@ -0,0 +1,202 @@
+/*
+ * Copyright (c) 2023, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <raft/core/device_mdarray.hpp>
+#include <raft/core/resource/thrust_policy.hpp>
+#include <raft/core/resources.hpp>
+#include <raft/linalg/map.cuh>
+#include <thrust/for_each.h>
+
+namespace raft::utils {
+/**
+ * @defgroup bitset Bitset
+ * @{
+ */
+/**
+ * @brief View of a RAFT Bitset.
+ *
+ * This lightweight structure stores a pointer to a bitset in device memory with it's length.
+ * It provides a test() device function to check if a given index is set in the bitset.
+ *
+ * @tparam IdxT Indexing type used. Default is uint32_t.
+ */
+template <typename IdxT = uint32_t>
+struct bitset_view {
+  using BitsetT            = uint32_t;
+  IdxT bitset_element_size = sizeof(BitsetT) * 8;
+
+  _RAFT_HOST_DEVICE bitset_view(BitsetT* bitset_ptr, IdxT bitset_len)
+    : bitset_ptr_{bitset_ptr}, bitset_len_{bitset_len}
+  {
+  }
+  /**
+   * @brief Create a bitset view from a device vector view of the bitset.
+   *
+   * @param bitset_span Device vector view of the bitset
+   */
+  _RAFT_HOST_DEVICE bitset_view(raft::device_vector_view<BitsetT, IdxT> bitset_span)
+    : bitset_ptr_{bitset_span.data_handle()}, bitset_len_{bitset_span.extent(0)}
+  {
+  }
+  /**
+   * @brief Device function to test if a given index is set in the bitset.
+   *
+   * @param sample_index Single index to test
+   * @return bool True if index has not been unset in the bitset
+   */
+  inline _RAFT_DEVICE auto test(const IdxT sample_index) const -> bool
+  {
+    const IdxT bit_element = bitset_ptr_[sample_index / bitset_element_size];
+    const IdxT bit_index   = sample_index % bitset_element_size;
+    const bool is_bit_set  = (bit_element & (1ULL << bit_index)) != 0;
+    return is_bit_set;
+  }
+  /**
+   * @brief Get the device pointer to the bitset.
+   */
+  inline _RAFT_HOST_DEVICE auto get_bitset_ptr() -> BitsetT* { return bitset_ptr_; }
```

Comment:
It might be nice to keep the name of this method more consistent with e.g. `std::vector` and similar (`.data()`) or `mdspan` and similar (`.data_handle()`).